core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1357,10 +1357,16 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
   ): ZIO[R1, E2, B] =
     new ZIO.Fold[R1, E, E2, A, B](
       self,
-      ZIOFn(() => that)(_.stripFailures match {
-        case None    => that
-        case Some(c) => ZIO.halt(c)
-      }),
+      ZIOFn(() => that) { cause =>
+        cause.stripFailures match {
+          case None =>
+            that.catchSomeCause {
```

Comment:
I think what you really want is `that.catchAll(cause2 => cause1 ++ cause2)`. There is no need to special case on die / failure (also they are not mutually exclusive, the same cause can be death and failure at the same time).

Reply:
If I just sum 2 causes, I would have a situation where `IO.fail(1) orElse IO.fail(2)` produce `ZIO.halt(Cause.die(FiberFailure(cause ++ cause2)))`, it's not possible to catch this with `either`, only with `catchAllCause`, and application dies.